### ISSUE TITLE
Bug: Multiple dex files define Lorg/powermock/api/mockito/expectation/WithExpectedArguments;

### ISSUE THREAD

**Issue by mikelupo** [powermock/powermock#821]
I have an written a small example Android app that is the "out of the box" single empty activity app.
When I try to add powermock via build.gradle, it will not run.
Below is my entire build.gradle file.

**Note:** Using compile because we're going to override getSystemService(String service) and embed some mocks within it to tweak the return(s) for ConnectivityManger and NetworkType.
```apply plugin: 'com.android.application'

android {
    compileSdkVersion 25
    buildToolsVersion "25.0.2"

    defaultConfig {
        applicationId "com.example.exampleapp"
        minSdkVersion 15
        targetSdkVersion 25
        versionCode 1
        versionName "1.0"

        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"

    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
}

def powerMockVersion =  '1.7.0'

dependencies {
    repositories {
        mavenCentral()
    }

    compile fileTree(dir: 'libs', include: ['*.jar'])

    compile ("org.powermock:powermock-module-junit4:$powerMockVersion") {
        exclude group: 'org.objenesis', module: 'objenesis'
    }

    compile ('org.powermock:powermock-api-mockito2:1.7.0') {
        exclude group: 'org.mockito', module: 'hamcrest-core'
        exclude group: 'org.hamcrest', module: 'hamcrest-core'
        exclude group: 'junit' exclude group: 'org.mockito'
        exclude group: 'org.objenesis', module: 'objenesis'
    }

//    androidTestCompile('com.android.support.test.espresso:espresso-core:2.2.2', {
//        exclude group: 'com.android.support', module: 'support-annotations'
//    })
//    testCompile 'junit:junit:4.12'
}
```

To reproduce:

- In your android project, create a new Android Module for an empty activity and call it exampleapp.
- Next, edit your build.gradle file for it. copy/paste the above.
- Gradle sync & Run it. 

Fails with the dexing error.
```
Error:Execution failed for task ':exampleapp:transformClassesWithDexForDebug'.
> com.android.build.api.transform.TransformException: com.android.ide.common.process.ProcessException: java.util.concurrent.ExecutionException: com.android.dex.DexException: Multiple dex files define Lorg/powermock/api/mockito/expectation/WithExpectedArguments;
```

Example project: http://23.79.234.224/Anaina/cache/TestData2/ficticious_project.tar

**Comment by mikelupo** [comment]
Hello @thekingnothing, I'm curious if any one knows of a workaround to this? We're currently blocked by this issue.

Thanks!

**Comment by thekingn0thing** [comment]
Unfortunately, I didn't have any experience with Android development. So I'm afraid that  cannot help you with the issue.

Best regrads,
Arthur Zagretdinov

On 25 Jul 2017, at 15:28, mikelupo <notifications@github.com<mailto:notifications@github.com>> wrote:


Hello @thekingnothing<https://github.com/thekingnothing>, I'm curious if any one knows of a workaround to this? We're currently blocked by this issue.

Thanks!

**Comment by mikelupo** [comment]
I found something of a work-around, but it leads to bigger problems in that Powermock is using java.beans rather than openbeans. 
[According to this StackOverflow post](https://stackoverflow.com/questions/15065545/using-jars-that-use-java-beans-classes-introspector-beaninfo-or-propertydescri#15185068)


Anyway, the workaround was to add the following to my android element of the build.gradle file for my test app.
dexOptions {
            jumboMode true
            multiDexEnabled true
        }

Now when calling spy(object to spy) I'm getting an exception thrown. 

```java.lang.IllegalStateException: Could not initialize plugin: interface org.mockito.plugins.MockMaker
at org.mockito.internal.configuration.plugins.PluginLoader$1.invoke(PluginLoader.java:66)
at java.lang.reflect.Proxy.invoke(Proxy.java:393)
at $Proxy0.isTypeMockable(Unknown Source)
at org.mockito.internal.util.MockUtil.typeMockabilityOf(MockUtil.java:29)
at org.mockito.internal.util.MockCreationValidator.validateType(MockCreationValidator.java:22)
at org.mockito.internal.creation.MockSettingsImpl.validatedSettings(MockSettingsImpl.java:186)
at org.mockito.internal.creation.MockSettingsImpl.confirm(MockSettingsImpl.java:180)
at org.mockito.internal.MockitoCore.mock(MockitoCore.java:62)
at org.mockito.Mockito.spy(Mockito.java:1812)
at com.akamai.android.sdk.internal.MapClientTestMainApp.getSystemService(MapClientTestMainApp.java:85)
at com.akamai.android.sdk.net.VocAccelerator.init(VocAccelerator.java:96)
at com.akamai.android.sdk.VocService.createVocService(VocService.java:143)
at com.akamai.vocsdkliveapitestlib.VocSdkLiveTest_BaseServiceTest.setVocService(VocSdkLiveTest_BaseServiceTest.java:217)
at com.akamai.vocsdkliveapitestlib.VocSdkLiveTest_BaseServiceTest.unRegisterIfNecessary(VocSdkLiveTest_BaseServiceTest.java:166)
at com.akamai.vocsdkliveapitestlib.VocSdkLiveTest_BaseServiceTest.waitForSyncToFinishAndThenDeregister(VocSdkLiveTest_BaseServiceTest.java:211)
at com.akamai.vocsdkliveapitestlib.VocSdkLiveTest_BaseServiceTest.unRegisterIfNecessary(VocSdkLiveTest_BaseServiceTest.java:151)
at com.akamai.vocsdkliveapitestlib.VocSdkLiveTest_BaseServiceTest.setUp(VocSdkLiveTest_BaseServiceTest.java:41)
at com.akamai.android.sdk.internal.MapLT_AICTests.setUp(MapLT_AICTests.java:64)
at junit.framework.TestCase.runBare(TestCase.java:132)
at junit.framework.TestResult$1.protect(TestResult.java:115)
at android.support.test.internal.runner.junit3.AndroidTestResult.runProtected(AndroidTestResult.java:77)
at junit.framework.TestResult.run(TestResult.java:118)
at android.support.test.internal.runner.junit3.AndroidTestResult.run(AndroidTestResult.java:55)
at junit.framework.TestCase.run(TestCase.java:124)
at android.support.test.internal.runner.junit3.NonLeakyTestSuite$NonLeakyTest.run(NonLeakyTestSuite.java:63)
at junit.framework.TestSuite.runTest(TestSuite.java:243)
at junit.framework.TestSuite.run(TestSuite.java:238)
at android.support.test.internal.runner.junit3.DelegatingTestSuite.run(DelegatingTestSuite.java:103)
at android.support.test.internal.runner.junit3.AndroidTestSuite.run(AndroidTestSuite.java:69)
at android.support.test.internal.runner.junit3.JUnit38ClassRunner.run(JUnit38ClassRunner.java:103)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at android.support.test.internal.runner.TestExecutor.execute(TestExecutor.java:59)
at android.support.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:262)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1970)
Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/beans/Introspector;
at org.powermock.configuration.support.ConfigurationMapper.map(ConfigurationMapper.java:45)
at org.powermock.configuration.support.ConfigurationReaderBuilder$ConfigurationReaderImpl.mapConfiguration(ConfigurationReaderBuilder.java:89)
at org.powermock.configuration.support.ConfigurationReaderBuilder$ConfigurationReaderImpl.createConfiguration(ConfigurationReaderBuilder.java:80)
at org.powermock.configuration.support.ConfigurationReaderBuilder$ConfigurationReaderImpl.read(ConfigurationReaderBuilder.java:71)
at org.powermock.configuration.support.ConfigurationFactoryImpl.readDefault(ConfigurationFactoryImpl.java:42)
at org.powermock.configuration.support.ConfigurationFactoryImpl.create(ConfigurationFactoryImpl.java:33)
at org.powermock.configuration.GlobalConfiguration.createConfig(GlobalConfiguration.java:75)
at org.powermock.configuration.GlobalConfiguration.<init>(GlobalConfiguration.java:59)
at org.powermock.configuration.GlobalConfiguration.mockitoConfiguration(GlobalConfiguration.java:40)
at org.powermock.api.mockito.mockmaker.PowerMockMaker.<init>(PowerMockMaker.java:45)
at java.lang.Class.newInstance(Native Method)
at org.mockito.internal.configuration.plugins.PluginLoader.loadImpl(PluginLoader.java:96)
at org.mockito.internal.configuration.plugins.PluginLoader.loadPlugin(PluginLoader.java:45)
at org.mockito.internal.configuration.plugins.PluginRegistry.<init>(PluginRegistry.java:18)
at org.mockito.internal.configuration.plugins.Plugins.<clinit>(Plugins.java:17)
at org.mockito.internal.configuration.plugins.Plugins.getMockMaker(Plugins.java:33)
at org.mockito.internal.util.MockUtil.<clinit>(MockUtil.java:24)
... 39 more
Caused by: java.lang.ClassNotFoundException: Didn't find class "java.beans.Introspector" on path: DexPathList[[zip file "/system/framework/android.test.runner.jar", zip file "/data/app/com.akamai.android.sdk.internal.test-1/base.apk", zip file "/data/app/com.akamai.android.sdk.internal-1/base.apk"],nativeLibraryDirectories=[/data/app/com.akamai.android.sdk.internal.test-1/lib/arm, /data/app/com.akamai.android.sdk.internal-1/lib/arm, /data/app/com.akamai.android.sdk.internal.test-1/base.apk!/lib/armeabi-v7a, /data/app/com.akamai.android.sdk.internal-1/base.apk!/lib/armeabi-v7a, /vendor/lib, /system/lib]]
at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
... 56 more
Suppressed: java.lang.ClassNotFoundException: java.beans.Introspector
at java.lang.Class.classForName(Native Method)
at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
... 57 more
Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available
```


I conclude that someone will have to fork this repo and perhaps make an android flavor.

**Comment by thekingn0thing** [comment]
@mikelupo,

Could you create a new issue and describe what is wrong with `java.beans` and how it could be fixed?

By the way, why do you want to package PowerMock into android application?

Best regrads,
Arthur Zagretdinov

On 25 Jul 2017, at 20:42, mikelupo <notifications@github.com<mailto:notifications@github.com>> wrote:


I found something of a work-around, but it leads to bigger problems in that Powermock is using java.beans rather than openbeans.
According to this StackOverflow post<https://stackoverflow.com/questions/15065545/using-jars-that-use-java-beans-classes-introspector-beaninfo-or-propertydescri#15185068>

Anyway, the workaround was to add the following to my android element of the build.gradle file for my test app.
dexOptions {
jumboMode true
multiDexEnabled true
}

Now when calling spy(object to spy) I'm getting an exception thrown.

at org.mockito.internal.configuration.plugins.PluginLoader$1.invoke(PluginLoader.java:66)
at java.lang.reflect.Proxy.invoke(Proxy.java:393)
at $Proxy0.isTypeMockable(Unknown Source)
at org.mockito.internal.util.MockUtil.typeMockabilityOf(MockUtil.java:29)
at org.mockito.internal.util.MockCreationValidator.validateType(MockCreationValidator.java:22)
at org.mockito.internal.creation.MockSettingsImpl.validatedSettings(MockSettingsImpl.java:186)
at org.mockito.internal.creation.MockSettingsImpl.confirm(MockSettingsImpl.java:180)
at org.mockito.internal.MockitoCore.mock(MockitoCore.java:62)
at org.mockito.Mockito.spy(Mockito.java:1812)
at com.akamai.android.sdk.internal.MapClientTestMainApp.getSystemService(MapClientTestMainApp.java:85)
at com.akamai.android.sdk.net.VocAccelerator.init(VocAccelerator.java:96)
at com.akamai.android.sdk.VocService.createVocService(VocService.java:143)
at com.akamai.vocsdkliveapitestlib.VocSdkLiveTest_BaseServiceTest.setVocService(VocSdkLiveTest_BaseServiceTest.java:217)
at com.akamai.vocsdkliveapitestlib.VocSdkLiveTest_BaseServiceTest.unRegisterIfNecessary(VocSdkLiveTest_BaseServiceTest.java:166)
at com.akamai.vocsdkliveapitestlib.VocSdkLiveTest_BaseServiceTest.waitForSyncToFinishAndThenDeregister(VocSdkLiveTest_BaseServiceTest.java:211)
at com.akamai.vocsdkliveapitestlib.VocSdkLiveTest_BaseServiceTest.unRegisterIfNecessary(VocSdkLiveTest_BaseServiceTest.java:151)
at com.akamai.vocsdkliveapitestlib.VocSdkLiveTest_BaseServiceTest.setUp(VocSdkLiveTest_BaseServiceTest.java:41)
at com.akamai.android.sdk.internal.MapLT_AICTests.setUp(MapLT_AICTests.java:64)
at junit.framework.TestCase.runBare(TestCase.java:132)
at junit.framework.TestResult$1.protect(TestResult.java:115)
at android.support.test.internal.runner.junit3.AndroidTestResult.runProtected(AndroidTestResult.java:77)
at junit.framework.TestResult.run(TestResult.java:118)
at android.support.test.internal.runner.junit3.AndroidTestResult.run(AndroidTestResult.java:55)
at junit.framework.TestCase.run(TestCase.java:124)
at android.support.test.internal.runner.junit3.NonLeakyTestSuite$NonLeakyTest.run(NonLeakyTestSuite.java:63)
at junit.framework.TestSuite.runTest(TestSuite.java:243)
at junit.framework.TestSuite.run(TestSuite.java:238)
at android.support.test.internal.runner.junit3.DelegatingTestSuite.run(DelegatingTestSuite.java:103)
at android.support.test.internal.runner.junit3.AndroidTestSuite.run(AndroidTestSuite.java:69)
at android.support.test.internal.runner.junit3.JUnit38ClassRunner.run(JUnit38ClassRunner.java:103)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at android.support.test.internal.runner.TestExecutor.execute(TestExecutor.java:59)
at android.support.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:262)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1970)
Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/beans/Introspector;
at org.powermock.configuration.support.ConfigurationMapper.map(ConfigurationMapper.java:45)
at org.powermock.configuration.support.ConfigurationReaderBuilder$ConfigurationReaderImpl.mapConfiguration(ConfigurationReaderBuilder.java:89)
at org.powermock.configuration.support.ConfigurationReaderBuilder$ConfigurationReaderImpl.createConfiguration(ConfigurationReaderBuilder.java:80)
at org.powermock.configuration.support.ConfigurationReaderBuilder$ConfigurationReaderImpl.read(ConfigurationReaderBuilder.java:71)
at org.powermock.configuration.support.ConfigurationFactoryImpl.readDefault(ConfigurationFactoryImpl.java:42)
at org.powermock.configuration.support.ConfigurationFactoryImpl.create(ConfigurationFactoryImpl.java:33)
at org.powermock.configuration.GlobalConfiguration.createConfig(GlobalConfiguration.java:75)
at org.powermock.configuration.GlobalConfiguration.<init>(GlobalConfiguration.java:59)
at org.powermock.configuration.GlobalConfiguration.mockitoConfiguration(GlobalConfiguration.java:40)
at org.powermock.api.mockito.mockmaker.PowerMockMaker.<init>(PowerMockMaker.java:45)
at java.lang.Class.newInstance(Native Method)
at org.mockito.internal.configuration.plugins.PluginLoader.loadImpl(PluginLoader.java:96)
at org.mockito.internal.configuration.plugins.PluginLoader.loadPlugin(PluginLoader.java:45)
at org.mockito.internal.configuration.plugins.PluginRegistry.<init>(PluginRegistry.java:18)
at org.mockito.internal.configuration.plugins.Plugins.<clinit>(Plugins.java:17)
at org.mockito.internal.configuration.plugins.Plugins.getMockMaker(Plugins.java:33)
at org.mockito.internal.util.MockUtil.<clinit>(MockUtil.java:24)
... 39 more
Caused by: java.lang.ClassNotFoundException: Didn't find class "java.beans.Introspector" on path: DexPathList[[zip file "/system/framework/android.test.runner.jar", zip file "/data/app/com.akamai.android.sdk.internal.test-1/base.apk", zip file "/data/app/com.akamai.android.sdk.internal-1/base.apk"],nativeLibraryDirectories=[/data/app/com.akamai.android.sdk.internal.test-1/lib/arm, /data/app/com.akamai.android.sdk.internal-1/lib/arm, /data/app/com.akamai.android.sdk.internal.test-1/base.apk!/lib/armeabi-v7a, /data/app/com.akamai.android.sdk.internal-1/base.apk!/lib/armeabi-v7a, /vendor/lib, /system/lib]]
at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
... 56 more
Suppressed: java.lang.ClassNotFoundException: java.beans.Introspector
at java.lang.Class.classForName(Native Method)
at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
... 57 more
Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available


I conclude that someone will have to fork this repo and perhaps make an android flavor.
